src/lib/storage/reference_segment.cpp

```diff
@@ -1,43 +1,45 @@
 #include "reference_segment.hpp"
 
 #include "storage/table.hpp"
+#include "types.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
 ReferenceSegment::ReferenceSegment(const std::shared_ptr<const Table>& referenced_table,
-                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos) {
-  // Implementation goes here
-}
+                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos)
+    : _referenced_table{referenced_table}, _referenced_column_id{referenced_column_id}, _pos_list{pos} {}
 
 AllTypeVariant ReferenceSegment::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return get_by_row_id(_pos_list->at(chunk_offset));
+}
+
+AllTypeVariant ReferenceSegment::get_by_row_id(RowID row_id) const {
+  if (row_id.is_null()) {
+    return NULL_VALUE;
+  }
+  const auto chunk = _referenced_table->get_chunk(row_id.chunk_id);
+  return (*chunk->get_segment(_referenced_column_id))[row_id.chunk_offset];
```

Comment:
Please do not use `operator[]()` (it is meant for debugging only).
Also consider using your accessor methods here.

Reply:
What would you suggest instead? We are in a function that is supposed to return `AllTypeVariant`, so we used the accessor that does just that, delegating the conversion from `T` to `AllTypeVariant` to whoever has the `T`.